python/ray/train/_internal/storage.py

```diff
@@ -405,10 +418,13 @@ def __init__(
         storage_path_provided = storage_path is not None
 
         self.storage_local_path = _get_defaults_results_dir()
-        # If `storage_path=None`, then set it to the local path.
+        # If `storage_path=None`, then set it to the ray artifact storage URI or
+        # the local path.
         # Invariant: (`storage_filesystem`, `storage_path`) is the location where
         # *all* results can be accessed.
-        self.storage_path = storage_path or self.storage_local_path
+        self.storage_path = (
+            storage_path or _get_artifact_storage_uri() or self.storage_local_path
```

Comment:
Should this be handled directly in the `RunConfig` instead? In my mind we would want to set this as early as possible, where Ray Storage is being read?

Reply:
We can do it in either location. Doing it here has the benefit that it's the path that every entrypoint goes through (e.g. both `Tuner()` and `tune.run()`).

But actually I think you're right, we should _also_ do it in the RunConfig.

Reply:
Ah I'm honestly a bit confused, maybe in that case it would make sense to centralize all this storage path evaluation logic into the `StorageContext`? cc @justinvyu

Reply:
I think we should leave the user's own `storage_path` alone in the `RunConfig`. I think it actually makes sense to do here -- all path handling / assigning defaults to be done in this one class.